bayes_opt/bayesian_optimization.py

```diff
@@ -260,21 +283,146 @@ def maximize(self,
+
+        x_probe: dict
+            point found by the optimization process
+        """
+
+        try:
+            x_array = numpy.array(list(x_probe.values()))
+
+        except AttributeError:
+            x_array = x_probe
+
+        min_distance = None
+        min_index = None
+        approximations = []
+
+        if self._target_column is None:
+            for row in dataset.itertuples():
+                dataset_tuple = []
+                for j in range(x_array.size):
+                    dataset_tuple.append(row[j + 1])
+
+                res = numpy.linalg.norm(x_array - dataset_tuple, 2)
+
+                if min_distance is None:
+                    min_distance = res
+                    approximations = self._space.array_to_params(dataset_tuple)
+                elif res == min_distance:
+                    approximations.append(self._space.array_to_params(dataset_tuple))
+                elif res < min_distance:
+                    min_distance = res
+                    approximations = self._space.array_to_params(dataset_tuple)
+            return random.choice(approximations)
+        else:
+            # dataset.drop(self._target_column)
+            for row in dataset.loc[:, dataset.columns != self._target_column].itertuples():
+                dataset_tuple = []
+                for j in range(x_array.size):
+                    dataset_tuple.append(row[j + 1])
+
+                res = numpy.linalg.norm(x_array - dataset_tuple, 2)
+
+                if min_distance.size == 0:
+                    min_index = row[0]
+                    min_distance = res
+                    approximations = self._space.array_to_params(dataset_tuple)
+                elif res == min_distance:
+                    if row[self._target_column] > dataset.iloc[min_index][self._target_column]:
+                        min_index = row[0]
+                        min_distance = res
+                        approximations = self._space.array_to_params(dataset_tuple)
+
+                elif res < min_distance:
+                    min_index = row[0]
+                    min_distance = res
+                    approximations = self._space.array_to_params(dataset_tuple)
+            return approximations, dataset.iloc[min_index][self._target_column]
+
+    def save_res_to_csv(self, is_approximation, exact_x=None):
+        """
+        A method to save results of the optimization to csv files located in results directory
+
+        Parameters
+        ----------
 
-            if self._bounds_transformer:
-                self.set_bounds(
-                    self._bounds_transformer.transform(self._space))
+        is_approximation: bool
+            true if the user passes a dataset as input
+        exact_x : list[dict]
+            contains exact x_probe
+        """
+        if is_approximation:
+            approximation_res = pd.DataFrame.from_dict(self.res)
+            approximation_res.to_csv("results/approximation.csv", index=True)
```

Comment:
I would call the file `approx_x.csv` instead of `approximation.csv`

Reply:
fixed